tower-middleware/Cargo.toml

```diff
@@ -0,0 +1,25 @@
+[package]
+name = "tower-middleware"
```

Comment:
are we convinced that "middleware" is the term we want to use? I'm still partial to "stack", but not enough to argue about it if there are folks who think middleware is better.

Reply:
🚲 🏠 😄 

I only reach for "middleware" because that is how its been referred to in prior ecosystems I've been involved with.

Pro: "middleware" has precedence. 
Con: It is long, possibly overloaded?

Stack feels like it refers to the stack of multiple middleware + service. Implementations of this trait are a "layer" of the stack. So, if anything `Layer` would be better... but that feels vague.

Reply:
I'm moderately pro-`Layer` personally. I think it's even reasonable to call the crate `tower-middleware` but the _trait_ `Layer` --- consider "a middleware layer which...."

Reply:
I guess for me personally I don't have strong feelings either way. I am just worried about our use of the term "middleware" as we use that in the readme. One thing to consider is what the API might look like for newer users, I think we should be careful how we use that terminology. 

### Current
```rust
ServiceBuilder::new(hyper)
        .middleware(InFlightLimitMiddleware::new(5))
        .middleware(RateLimitMiddleware::new(5, Duration::from_secs(1)))
        .build(dst);
```

### Layer
```rust
ServiceBuilder::new(hyper)
        .push(InFlightLimitLayer::new(5))
        .push(RateLimitLayer::new(5, Duration::from_secs(1)))
        .build(dst);
```

I do kind of like the name layer as its a bit more clear and differentiates between the high-level goal that tower is trying to provide and the direct goal you are trying to achieve by _layering_ the middleware.

And in that sense too, I would rather the crate be called `tower-layer` to make it a bit more clear. I would imagine someone reading the readme and seeing half the crate descriptions start with "Middleware that..." then seeing a `tower-middleware` might get a bit confused.

Plus, `Layer<S, Request>` is a lot less typing 😅 .

Reply:
I'm +1 on Layer. Coincidentally, this is what I chose for a similar trait in linkerd2-proxy.

Reply:
I'm fine w/ `Layer` as well. It fits well.

Reply:
The crate *probably* should be renamed `tower-stack` at that point? So, a "stack" is a set of layers w/ a service.

Reply:
A benefit of a stack/layer naming convention is that, historically, “middleware” was primarily used in server contexts. Since Tower's services can be used in a client _or_ server context, a new naming convention for a new concept is fitting.

Reply:
On my way out the door, but my 2c: I think the name should communicate what the property of this "thing" is that differentiates it from just any ol' `Service`. And I _believe_ that property is that it's a 1-1 mapping. I personally believe "layer" doesn't carry that implication, but will defer to consensus. I like words like "wrap", "chain", and "link" for this.